04-onion-routing.md

```diff
@@ -239,6 +239,49 @@ When forwarding HTLCs, nodes MUST construct the outgoing HTLC as specified withi
 `per_hop` above; otherwise, deviation from the specified HTLC parameters
 may lead to extraneous routing failure.
 
+## Non-strict Forwarding
+
+A node MAY forward an HTLC along an outgoing channel other than the one
+specified by `short_channel_id`, so long as the receiver has the same node
+public key intended by `short_channel_id`. Thus, if `short_channel_id` connects
+nodes A and B, the HTLC can forwarded across any channel connecting A and B.
+Failure to adhere will result in the receiver being unable to decrypt the next
+hop in the onion packet.
+
+### Rationale
+
+In the event that two peers have multiple channels, the downstream node will be
+able to decrypt the next hop payload regardless of which channel the packet is
+sent across.
+
+Nodes implementing non-strict forwarding are able to make real-time assessments
+of channel bandwidths with a particular peer, and use the channel that is
+locally-optimal. 
+
+For example, if the channel specified by `short_channel_id` connecting A and B
+does not have enough bandwidth at forwarding time, then A is able use a
+different channel that does. This can reduce payment latency by preventing the
+HTLC from failing due to bandwidth constraints across `short_channel_id`, only
+to have the sender attempt the same route differing only in the channel between
+A and B.
+
+Non-strict forwarding allows nodes to make use of private channels connecting
+them to the receiving node, even if the channel is not known in the public
+channel graph.
+
+### Recommendation
+
+Implementations using non-strict forwarding should consider applying the same
```

Comment:
I think we should also recommend sending errors for the initially requested `short_channel_id` (at least until we allow sending multiple `channel_update`). This makes it transparent for the sender: either the payment succeeds using a different channel and it isn't even aware of what happened, or there is an error and it gets an error for the channel it asked.

There are other corner cases, like what if we return an error for an unannounced channel?

The way we implemented it in eclair is that we may try different channels, but always end with the one that was requested and return that last error.

Reply:
I think it can be worth distinguishing between three cases where non-strict forwarding fails:
1. The htlc does not satisfy the requested channel's policy.
2. The htlc does satisfy the requested channel's policy (the payment failed for a different reason like insufficient balance) and there is no other channel to B that would be able to carry the payment at this time (for example when none of the channels has enough balance for the requested amount).
3. The htlc does satisfy the requested channel's policy and there are one or more other channels to B that would be able to carry the payment at this time if their policy would have been satisfied.

In case 1, the sender's graph is not up to date. It is probably best to return a `channel_update` for the requested channel so the sender can try again.

In case 2, returning a `channel_update` for the requested channel is unnecessary, because the sender is already aware of the current policy. If optional in the failure message, leave it out to save bytes.

In case 3, returning a `channel_update` for the requested channel is unnecessary, because the sender is already aware of the current policy. But we could choose to re-purpose the `channel_update` to signal a policy that would allow forwarding (assuming that nothing changes in the mean time). 

The forwarding node could iterate over all its channels to B that have enough balance for the requested amount and determine the best channel for the sender. "Best" is based on a tradeoff between fee and timelock, similar to what is used during path finding. A `channel_update` is returned for the best channel. This behaviour could be signaled with a node feature bit, so that senders know how to interpret a `channel_update` for a channel other than the requested channel. It allows them to find a successful route faster and increases the chance for the forwarding node to earn a fee.

A forwarding node could also try to earn more by signaling a more expensive channel, but it may be that the sender then first tries other routes, potentially leaving the forwarding node with empty hands.

In case the best channel is a private channel, the forwarding node could send back a `channel_update` for a public channel that has a policy that is 'compatible' (equal or higher fees and/or cltv delta) with the private channel. It doesn't matter if this public channel doesn't have enough balance, because non-strict forwarding will also be applied on the next attempt.

Reply:
> If optional in the failure message, leave it out to save bytes.

It wouldn't really save bytes because the data is [padded to 256B anyway](https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md#requirements-1).

> But we could choose to re-purpose the channel_update to signal a policy that would allow forwarding (assuming that nothing changes in the mean time).

A proposal made during the summit was that we should allow returning multiple `channel_update`. The more I think about it though, the more I'm convinced that if supporting non-strict forwarding, a per-node policy is the only thing that really makes sense. That simply removes your 3rd point and weird corner cases (private channels and incentive issues). Anyway I think @cfromknecht intended to specify shorter term recommendations, without changing the onion's format.

Reply:
Point taken on padding.

I can also see this all going towards a per-node policy. Indeed, the goal of the 3rd point is a short term improvement to the current behaviour without changing the onion format.